include/kernel.h

```diff
@@ -0,0 +1,534 @@
+/**
+ * @file kernel.h
+ * @brief Multi-threading kernel
+ * @ingroup kernel
+ */
+#ifndef KERNEL_H
+#define KERNEL_H
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "exception.h"
+
+/**
+ * @defgroup kernel Multi-threading kernel
+ * @ingroup libdragon
+ * @brief A multi-threading kernel that can be used by libdragon applications.
+ *
+ * This module implements a hybrid cooperative/preemptive multi-threaded
+ * kernel for parallel execution of code.
+ *
+ * The kernel is "hard real-time": any thread with high priority that is
+ * ready (not blocked) will always have priority over lower priority threads.
+ * This is required to allow threads to implement operations that require
+ * low latency, for instance preparing audio when the AI interrupt fires.
+ *
+ * The kernel is not fully preemptive; in particular, there is no timer
+ * interrupt that switches among ready threads at a fixed interval. This
+ * is not deemed necessary as applications will usually have a low number
+ * threads that are mostly blocked waiting for specific events like
+ * hardware interrupts or background activities like RSP ucode.
+ *
+ * An event, represented by a #kevent_t instance, is a signal that can
+ * be broadcasted to many different threads. A thread can use
+ * #kthread_wait_event to wait for a specific event to fire. The kernel
+ * library includes a set of events bound to MI peripheral interrupts,
+ * so that it's possible for threads to wait for specific interrupts and
+ * act after them.
+ *
+ * Typically, applications will keep the main thread at low priority
+ * for the main business logic, and will spawn a few threads to handle
+ * high-priority / low-latency responses to hardware events. Libdragon
+ * itself offers libraries that need threading to provide a easy-to-use
+ * API.
+ *
+ * A context-switch between threads can happen in the following situations:
+ *
+ *   * Preemptively: when a hardware interrupt generates a event that
+ *     wakes up a thread (see #kevent). This is the most common scenario:
+ *     an event connected to an interrupt is fired, and a higher priority
+ *     thread waiting for the event is woken up.
+ * 
+ *   * Cooperatively: when a thread explicitly calls a blocking function
+ *     like #kthread_sleep or #kthread_wait_event to wait for an event.
+ *     The kernel will put the current thread to sleep and schedule the
+ *     highest-priority thread among the ready ones, including threads
+ *     that have a priority lower than the one that goes to sleep. When
+ *     the event is fired, the lower-priority thread will be preempted
+ *     to give back control to the thread that just woke up.
```

Comment:
So we always give control to a just woken up thread regardless of its priority right? What should happen when there are more than one thread need to wake up to the same event (say A and B)? From what I understand from the general layout, the one with the highest priority should wake up (let's say thread A). Then thread B would not get scheduled if there are higher priority threads that will get scheduled if thread A cooperatively yileds. This might not be a practical case but I'm speculating to understand/predict the edge cases.

Reply:
> So we always give control to a just woken up thread regardless of its priority right?

I think the documentation might be confusing.

Let's says you have three ready threads A (priority 5), B (priority 3), C (priority 3). A is running by definition (since a higher-priority ready thread is always scheduled, winning against lower priority threads). If A explicitly goes to sleep (with either kthread_sleep or kthread_wait_event), at that point the scheduler will schedule a lower priority thread, so either B or C. Notice that if A goes to sleep, *by definition* a lower-or-equal priority thread is scheduled; it's impossible that a higher priority thread is scheduled because if there was a higher priority thread D ready while A was running, A would have been stopped immediately to run D.

Let me know if now it's more clear and how can we make the documentation more clear (though keeping it concise of course).

> What should happen when there are more than one thread need to wake up to the same event (say A and B)? From what I understand from the general layout, the one with the highest priority should wake up (let's say thread A).

All threads are waken up, they all become ready (both A and B). Then the scheduler gives CPU to the currently ready thread that has the highest priority. This could be A like your in example or maybe it could be C (the thread that triggered the event) that might have a higher priority than both A and B. In this case, C keeps running after triggering the event, but A and B are both ready; as soon as C goes to sleep for any reason, they might be scheduled.

Reply:
Maybe we can state the rules of the threading system somewhere and verify the documentation makes sense/is consistent considering the rules. Most of this is already stated in the docs, but having them like this might help;

- There is always an active thread that is currently running.
- When the active thread goes to sleep, the library switches to the thread with the highest priority from the `ready` pool.
- The active thread will go to sleep only when it yields via a `kthread_` call in its entry point function.

I'm not sure if this is complete yet but would help a lot understanding it, at least for me.

Then it is the consumer's responsibility to make sure a thread will eventually get shceduled etc.

Reply:
I've updated the documentation. Hopefully is more clear now. Please let me know.